apiserver/facades/client/secrets/secrets.go

```diff
@@ -338,7 +343,89 @@ func fromUpsertParams(p params.UpsertSecretArg) state.UpdateSecretParams {
 	}
 }
 
-// CreateSecrets isn't on the v1 API.
+// UpdateSecrets isn't on the v1 API.
+func (s *SecretsAPIV1) UpdateSecrets(_ struct{}) {}
+
+// UpdateSecrets creates new secrets.
+func (s *SecretsAPI) UpdateSecrets(args params.UpdateUserSecretArgs) (params.ErrorResults, error) {
+	result := params.ErrorResults{
+		Results: make([]params.ErrorResult, len(args.Args)),
+	}
+	if err := s.checkCanAdmin(); err != nil {
+		return result, errors.Trace(err)
+	}
+	backend, err := s.getBackend(nil)
+	if err != nil {
+		return result, errors.Trace(err)
+	}
+	for i, arg := range args.Args {
+		err := s.updateSecret(backend, arg)
+		if errors.Is(err, state.LabelExists) {
+			err = errors.AlreadyExistsf("secret with label %q", *arg.Label)
+		}
+		result.Results[i].Error = apiservererrors.ServerError(err)
+	}
+	return result, nil
+}
+
+func (s *SecretsAPI) updateSecret(backend provider.SecretsBackend, arg params.UpdateUserSecretArg) (errOut error) {
+	if err := arg.Validate(); err != nil {
+		return errors.Trace(err)
+	}
+	uri, err := coresecrets.ParseURI(arg.URI)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	md, err := s.secretsState.GetSecret(uri)
+	if err != nil {
+		// Check if the uri exists or not.
+		return errors.Trace(err)
+	}
+	if len(arg.Content.Data) > 0 {
+		revId, err := backend.SaveContent(context.TODO(), uri, md.LatestRevision+1, coresecrets.NewSecretValue(arg.Content.Data))
+		if err != nil && !errors.Is(err, errors.NotSupported) {
+			return errors.Trace(err)
+		}
+		if err == nil {
+			defer func() {
+				if errOut != nil {
+					// If we failed to update the secret, we should delete the
+					// secret value from the backend for the new revision.
+					if err2 := backend.DeleteContent(context.TODO(), revId); err2 != nil &&
+						!errors.Is(err2, errors.NotSupported) &&
+						!errors.Is(err2, errors.NotFound) {
+						logger.Errorf("failed to delete secret %q: %v", revId, err2)
+					}
+				}
+			}()
+			arg.Content.Data = nil
+			arg.Content.ValueRef = &params.SecretValueRef{
+				BackendID:  s.activeBackendID,
+				RevisionID: revId,
+			}
+		}
+	}
+	_, err = s.secretsState.UpdateSecret(uri, fromUpsertParams(arg.UpsertSecretArg))
+	return errors.Trace(err)
+}
+
+// RemoveSecrets isn't on the v1 API.
+func (s *SecretsAPIV1) RemoveSecrets(_ struct{}) {}
+
+// RemoveSecrets remove user secret.
+func (s *SecretsAPI) RemoveSecrets(args params.DeleteSecretArgs) (params.ErrorResults, error) {
+	return commonsecrets.RemoveSecretsForClient(
+		s.secretsState, s.backendConfigGetter,
+		s.authTag, args,
+		func(*coresecrets.URI) error {
+			// Admins can delete any secret in the model.
+			return errors.Trace(s.checkCanAdmin())
```

Comment:
Users should only be able to delete users created secrets, never charm created secrets. So just as there is a "canManage" check for charm secrets, the equivalent check for users secrets is to see if the owner is the model. The only exception would be if for some reason the owner app or unit does not exist any more, but that cannot happen since we delete secrets as the app is removed etc.